### PR TITLE
Pattern detected in webhooks message #560

### LINKED ISSUE #560: Pattern detected in webhooks message
Now we get something like
```
Analytic unit: NGINX_PROD_REGULAR_PEAK
From: Thu Apr 11 2019 18:26:10 GMT+0000 (UTC)
To: Thu Apr 11 2019 18:28:50 GMT+0000 (UTC)
```

But let's do it like:
```
[PATTERN DETECTED]: NGINX_PROD_REGULAR_PEAK
From: Thu Apr 11 2019 18:26:10 GMT+0000 (UTC)
To: Thu Apr 11 2019 18:28:50 GMT+0000 (UTC)
```

so it's similar to 

```
[Alerting] CPU Usage 70%
URL: http://grafana.com/grafana/d/MdMbm-qik/prometheus?fullscreen=true&edit=true&tab=alert&panelId=278&orgId=1

Metrics:
cpu_usage: 1.494
```

### PR DESCRIPTION
**Pull request by amper43**
fixes #560 

## Changes
- deduplicate datasource unavailable/available webhooks

### CODE REVIEW

File: analytics/analytics/server.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
 
     if res['status'] == 'SUCCESS' and res['payload'] is not None:
         res['_id'] = task['_id']
-        message = services.server_service.ServerMessage('DETECT', res)
+        message = services.server_service.ServerMessage('PUSHDETECT', res)
```
Comment (rozetko) [inline analytics/analytics/server.py]: Maybe `PUSH_DETECT` / `PUSH_DETECTION`?

File: server/src/services/alert_service.ts [hunk 2]
```diff
@@ -110,33 +112,33 @@ export class AlertService {
   public getGrafanaAvailableReporter() {
     if(!this._grafanaAvailableReporter) {
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
    if(this._grafanaAvailableReporter === undefined) {
```

File: server/src/services/alert_service.ts [hunk 3]
```diff
@@ -110,33 +112,33 @@ export class AlertService {
...
       this._grafanaAvailableReporter = availableReporter(
```
Comment (rozetko) [inline server/src/services/alert_service.ts]: Is `availableReporter` a class?

File: server/src/services/alert_service.ts [hunk 4]
```diff
@@ -110,33 +112,33 @@ export class AlertService {
...
   }
 
-  public getAvailableWebhook(recoveryMsg: string, failureMsg: string) {
-    return availableReporter(
-      [recoveryMsg, WebhookType.RECOVERY],
-      [failureMsg, WebhookType.FAILURE],
-      this.sendMsg,
-      this.sendMsg
-    );
+  public datasourceAvailableWebhook(url: string) {
```
Comment (rozetko) [inline server/src/services/alert_service.ts]: `datasourceAvailableWebhook` -> `sendDatasourceAvailableWebhook`

File: server/src/services/alert_service.ts [hunk 5]
```diff
@@ -110,33 +112,33 @@ export class AlertService {
...
+    const reporter = this._getDatasourceAvailableReporter(url);
+    reporter(true);
+  }
+
+  public datasourceUnavailableWebhook(url: string) {
```
Comment (rozetko) [inline server/src/services/alert_service.ts]: `datasourceUnavailableWebhook` -> `sendDatasourceUnavailableWebhook`

File: server/src/services/alert_service.ts (hunk 3 above)
Comment (amper43) [reply server/src/services/alert_service.ts]: @rozetko this is function

File: server/src/services/alert_service.ts [hunk 6]
```diff
@@ -73,12 +74,21 @@ class ThresholdAlert extends Alert {
 
 export class AlertService {
 
-  private _alerts: { [id: string]: Alert; };
+  private _alerts: { [id: string]: Alert };
   private _alertingEnable: boolean;
   private _grafanaAvailableReporter: Function;
+  private _datasourceAvailableReporters: { [url: string]: Function };
```
Comment (amper43) [inline server/src/services/alert_service.ts]: use `Map` here

File: server/src/services/alert_service.ts (hunk 3 above)
Comment (jonyrock) [reply server/src/services/alert_service.ts]: @amper43 when you should have a verb in this function name. It looks like a class name.

File: server/src/services/alert_service.ts [hunk 7]
```diff
@@ -1,34 +1,54 @@
...
-      sendAnalyticWebhook(alert);
+      sendNotification(this.makeNotification(segment));
     }
   };
 
-  protected makeAlert(segment): AnalyticAlert {
-    const alert: AnalyticAlert = {
+  protected makeNotification(segment: Segment): Notification {
+    const meta = this.makeMeta(segment);
+    const message =this.makeMessage(meta);
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
    const message = this.makeMessage(meta);
```

File: server/src/services/alert_service.ts [hunk 8]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
       if(this.enabled) {
-        sendAnalyticWebhook(this.makeAlert(segment));
+        sendNotification(this.makeNotification(segment));
       }
     }
   }
+
+  protected makeMessage(meta: AnalyticMeta): string {
+    return [
+    `[PATTERN DETECTED] ${meta.analyticUnitName}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `[PATTERN DETECTED] ${meta.analyticUnitName}`,
```

File: server/src/services/alert_service.ts [hunk 9]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `URL: ${meta.grafanaUrl}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `URL: ${meta.grafanaUrl}`,
```

File: server/src/services/alert_service.ts [hunk 10]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    ``,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      ``,
```

File: server/src/services/alert_service.ts [hunk 11]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `From: ${new Date(meta.from)}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `From: ${new Date(meta.from)}`,
```

File: server/src/services/alert_service.ts [hunk 12]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `To: ${new Date(meta.to)}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `To: ${new Date(meta.to)}`,
```

File: server/src/services/alert_service.ts [hunk 13]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `ID: ${meta.analyticUnitId}`
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `ID: ${meta.analyticUnitId}`
```

File: server/src/services/alert_service.ts [hunk 14]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
         }
       }
 
       this.lastOccurence = segment.from;
     }
   }
+
+  protected makeMessage(meta: AnalyticMeta): string {
+    let message = [
+    `[TRESHOLD ALERTING] ${meta.analyticUnitName}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `[THRESHOLD ALERTING] ${meta.analyticUnitName}`,
```

File: server/src/services/alert_service.ts [hunk 15]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `URL: ${meta.grafanaUrl}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `URL: ${meta.grafanaUrl}`,
```

File: server/src/services/alert_service.ts [hunk 16]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    ``,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      ``,
```

File: server/src/services/alert_service.ts [hunk 17]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `Starts at: ${new Date(meta.from)}`,
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `Starts at: ${new Date(meta.from)}`,
```

File: server/src/services/alert_service.ts [hunk 18]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    `ID: ${meta.analyticUnitId}`
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
      `ID: ${meta.analyticUnitId}`
```

File: server/src/services/alert_service.ts [hunk 19]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+      `To: ${new Date(meta.to)}`,
+      `ID: ${meta.analyticUnitId}`
+    ].join('\n');
+  }
 };
 
 
 class ThresholdAlert extends Alert {
+  // TODO: configure threshold timing in panel like Grafana's alerts (`evaluate` time, `for` time)
+  // TODO: make events for starn and end of treshold
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
  // TODO: make events for start and end of threshold
```

File: server/src/services/alert_service.ts [hunk 20]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+      `[THRESHOLD ALERTING] ${meta.analyticUnitName}`,
+      `URL: ${meta.grafanaUrl}`,
+      ``,
+      `Starts at: ${new Date(meta.from)}`,
+      `ID: ${meta.analyticUnitId}`
+    ].join('\n');
+
+    if(meta.params !== undefined) {
+      const metrics = `
+      Metrics:
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
        Metrics:
```

File: server/src/services/alert_service.ts [hunk 21]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+      ${this.analyticUnit.metric.targets[0].expr}: ${meta.params.value}
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
        ${this.analyticUnit.metric.targets[0].expr}: ${meta.params.value}
```

File: server/src/services/alert_service.ts [hunk 22]
```diff
@@ -39,46 +59,87 @@ class PatternAlert extends Alert {
...
+    }
+    return message;
+  }
 }
 
 
 export class AlertService {
 
-  private _alerts: { [id: string]: Alert; };
+  private _alerts: { [id: string]: Alert };
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
  TODO: object -> Map
  private _alerts: { [id: string]: Alert };
```

File: server/src/services/alert_service.ts [hunk 23]
```diff
@@ -39,46 +59,88 @@ class PatternAlert extends Alert {
...
+    }
+    return message;
+  }
 }
 
 
 export class AlertService {
 
-  private _alerts: { [id: string]: Alert; };
+  TODO: object -> Map
```
Comment (rozetko) [inline server/src/services/alert_service.ts]:
```suggestion
  // TODO: object -> Map
```

**Comment by jonyrock** [comment]
Please provide description about the change. I want to show an example.

**Pull request merged**